src/utils.py

```diff
@@ -1394,6 +1394,122 @@ def run_hill_climbing(
+        self._collision_fn = collision_fn
+        self._distance_fn = distance_fn
+        self._rng = rng
+        self._num_attempts = num_attempts
+        self._num_iters = num_iters
+        self._smooth_amt = smooth_amt
+
+    def query(self, pt1: _S, pt2: _S) -> Optional[List[_S]]:
+        """Query the BiRRT, to get a collision-free path from pt1 to pt2.
+
+        If none is found, returns None.
+        """
+        if self._collision_fn(pt1) or self._collision_fn(pt2):
+            return None
+        direct_path = self._try_direct_path(pt1, pt2)
+        if direct_path is not None:
+            return direct_path
+        for _ in range(self._num_attempts):
+            path = self._rrt_connect(pt1, pt2)
+            if path is not None:
+                return self._smooth_path(path)
+        return None
+
+    def _try_direct_path(self, pt1: _S, pt2: _S) -> Optional[List[_S]]:
+        path = [pt1]
+        for newpt in self._extend_fn(pt1, pt2):
+            if self._collision_fn(newpt):
+                return None
+            path.append(newpt)
+        return path
+
+    def _rrt_connect(self, pt1: _S, pt2: _S) -> Optional[List[_S]]:
+        root1, root2 = _BiRRTNode(pt1), _BiRRTNode(pt2)
+        nodes1, nodes2 = [root1], [root2]
+
+        def _get_pt_dist_to_node(pt: _S, node: _BiRRTNode[_S]) -> float:
+            return self._distance_fn(pt, node.data)
+
+        for _ in range(self._num_iters):
+            if len(nodes1) > len(nodes2):
+                nodes1, nodes2 = nodes2, nodes1
+            samp = self._sample_fn(pt1)
+            min_key1 = functools.partial(_get_pt_dist_to_node, samp)
+            nearest1 = min(nodes1, key=min_key1)
+            for newpt in self._extend_fn(nearest1.data, samp):
+                if self._collision_fn(newpt):
+                    break
+                nearest1 = _BiRRTNode(newpt, parent=nearest1)
+                nodes1.append(nearest1)
+            min_key2 = functools.partial(_get_pt_dist_to_node, nearest1.data)
+            nearest2 = min(nodes2, key=min_key2)
+            for newpt in self._extend_fn(nearest2.data, nearest1.data):
+                if self._collision_fn(newpt):
+                    break
+                nearest2 = _BiRRTNode(newpt, parent=nearest2)
+                nodes2.append(nearest2)
+            else:
+                path1 = nearest1.path_from_root()
+                path2 = nearest2.path_from_root()
+                # This is a tricky case to cover.
+                if path1[0] != root1:  # pragma: no cover
+                    path1, path2 = path2, path1
+                assert path1[0] == root1
+                path = path1[:-1] + path2[::-1]
+                return [node.data for node in path]
+        return None
+
+    def _smooth_path(self, path: List[_S]) -> List[_S]:
+        # This is a tricky case to cover and may not ever happen in practice,
+        # but leaving it here just in case.
+        if len(path) <= 2:  # pragma: no cover
```

Comment:
i don't get how it can happen in practice... if the path length is 0 or 1, wouldn't `try_direct_path` have worked?

should we just make this an assert?